### PR TITLE
Add boiler-plate main.c file for building stdc test executables.

### PR DESCRIPTION
**Pull request by simonqhughes**
The following provides more information on this commit:
- In order for the stdc target to build an executable test binary this main.c
  is required to call the val_entry() function. It can be used to build CRYPTO,
  PROTECTED_STORAGE and INTERNAL_TRUSTED_STORAGE test executables, for example.

### CODE REVIEW

**Comment by simonqhughes** [comment]
This PR addresses item 2 in the following email from @jaypit02 

```
From: Jay 
Sent: Friday, September 20, 2019 2:15 PM
Subject: mbed-linux target support

Hi Mbed-linux team,

This is the follow up mail on following items that we discussed in previous meeting.
<lines deleted>

2.	Binary output support- Add main app compilation and Link step
o	We are investigating various options to support this. We’ll let you know once we conclude the final option. Presume, it will be post-compilation option to the build script such that user can perform compilation of the main app and linkage of required objects. Again, all of these can part of target tgt_dev_apis_stdc directory.

<lines deleted>

Please feel to share your comments and suggestions on these points.

Thanks,
Jay
```

The first step towards getting support for building the binaries is having the main.c as submitted in this PR. Please can you accept it?

The second step is having some psa-arch-test make/cmake support to build main.o. The psa_trusted_storage_linux has the following makefile to build psa-arch-test PSA storage test binaries:

https://github.com/ARMmbed/psa_trusted_storage_linux/blob/master/test/Makefile

in particular the following lines:

```
###############################################################################
# PSA Compliance psa-arch-tests-ps protected storage test binary targets
###############################################################################
# Target to build psa-arch-tests protected storage test binary
.PHONY: psa-arch-tests-ps
psa-arch-tests-ps: mbed-crypto psa-storage-install $(TOPDIR)/psa-arch-tests/api-tests/build-ps $(TOPDIR)/psa-arch-tests/api-tests/build-ps/CMakeFiles main.o
	cd psa-arch-tests/api-tests/build-ps && cmake --build . -- VERBOSE=$(PSA_ARCH_TESTS_CMAKE_VERBOSE)
	cd psa-arch-tests/api-tests/build-ps && gcc -o psa-arch-tests-ps $(TOPDIR)/$(PSA_ARCH_TESTS_DIR)/$(PSA_ARCH_TESTS_MAIN_C_PATH)/main.o ./dev_apis/protected_storage/test_combine.a ./val/val_nspe.a ./platform/pal_nspe.a ./dev_apis/protected_storage/test_combine.a $(TOPDIR)/$(PSA_STORAGE_DIR)/usr/local/lib/libpsastorage.so

# Target to create psa-arch-tests protected storage build directory
$(TOPDIR)/psa-arch-tests/api-tests/build-ps:
	mkdir $(TOPDIR)/psa-arch-tests/api-tests/build-ps

# Target to invoke psa-arch-tests protected storage cmake configuration.
$(TOPDIR)/psa-arch-tests/api-tests/build-ps/CMakeFiles:
	cd psa-arch-tests/api-tests/build-ps && cmake ../ -DTOOLCHAIN=HOST_GCC -DTARGET=tgt_dev_apis_stdc -DSUITE=PROTECTED_STORAGE -DPSA_INCLUDE_PATHS=$(TOPDIR)/$(PSA_STORAGE_DIR)/inc -DCPU_ARCH=armv7m

# Target to psa-arch-tests protected storage test binary build
.PHONY: psa-arch-tests-ps-build
psa-arch-tests-ps-build:
	cd $(TOPDIR)/psa-arch-tests/api-tests/build-ps && cmake --build .

# Target to delete the psa-arch-tests-ps build directory and other artifacts
.PHONY: psa-arch-tests-ps-clean
psa-arch-tests-ps-clean:
	rm -fR $(TOPDIR)/psa-arch-tests/api-tests/build-ps
	rm -fR $(TOPDIR)/$(PSA_ARCH_TESTS_DIR)/$(PSA_ARCH_TESTS_MAIN_C_PATH)/main.o

# Target to compile boiler-plate main.c for linking with psa-arch-tests libraries.
main.o:
	gcc -Wall -Werror -c -o $(TOPDIR)/$(PSA_ARCH_TESTS_DIR)/$(PSA_ARCH_TESTS_MAIN_C_PATH)/main.o $(TOPDIR)/$(PSA_ARCH_TESTS_DIR)/$(PSA_ARCH_TESTS_MAIN_C_PATH)/main.c

# Target to run psa-arch-tests protected storage test binary.
.PHONY: test-psa-arch-tests-ps
test-psa-arch-tests-ps: $(TEST_DIR) psa-arch-tests-ps
	-$(TOPDIR)/psa-arch-tests/api-tests/build-ps/psa-arch-tests-ps

```

in summary, we need you to decide how you want the following target to be incorporated into the code base:

```
# Target to compile boiler-plate main.c for linking with psa-arch-tests libraries.
main.o:
	gcc -Wall -Werror -c -o $(TOPDIR)/$(PSA_ARCH_TESTS_DIR)/$(PSA_ARCH_TESTS_MAIN_C_PATH)/main.o $(TOPDIR)/$(PSA_ARCH_TESTS_DIR)/$(PSA_ARCH_TESTS_MAIN_C_PATH)/main.c
```

**Comment by simonqhughes** [comment]
This PR now has working target.cmake changes for building PSA storage test binaries.

File: api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake [hunk 1]
```diff
@@ -15,6 +15,53 @@
...
+
+	# Create list of test binary source files.
+	list(APPEND EXE_SRC ${PSA_ROOT_DIR}/platform/targets/${TARGET}/nspe/common/main.c)
+
+	# Create list of libraries to link to test binary
+	list(APPEND EXE_LIBS
+		${PROJECT_BINARY_DIR}/dev_apis/${_api_dir}/test_combine.a
+		${PROJECT_BINARY_DIR}/val/val_nspe.a
+		${PROJECT_BINARY_DIR}/platform/pal_nspe.a
+		${PROJECT_BINARY_DIR}/dev_apis/${_api_dir}/test_combine.a
```
Comment (jaypit02) [inline api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake]: test_combine.a included twice. Please remove this instance.

File: api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake [hunk 2]
```diff
@@ -44,10 +90,16 @@ if(${SUITE} STREQUAL "CRYPTO")
 	)
 endif()
 if(${SUITE} STREQUAL "PROTECTED_STORAGE")
-	message(FATAL_ERROR "Protected Storage not supported")
+	list(APPEND PAL_SRC_C_NSPE
+		${PSA_ROOT_DIR}/platform/targets/${TARGET}/nspe/protected_storage/pal_protected_storage_intf.c
+	)
+	_create_psa_storage_exe(psa-arch-tests-ps protected_storage)
```
Comment (jaypit02) [inline api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake]: Can similar binary support for crypto added as part of this PR?

File: api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake (hunk 1 above)
Comment (simonqhughes) [reply api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake]: Removed as requested.

File: api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake (hunk 2 above)
Comment (simonqhughes) [reply api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake]: Added as requested.

**Comment by simonqhughes** [comment]
@jaypit02 @kotegowder please can you review updated commits and approve? Thanks.

File: api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake [hunk 3]
```diff
@@ -15,6 +15,61 @@
...
+#   _exe_name     Name of the test binary to generate.
+#   _api_dir      PSA API directory name e.g. crypto, 
+#                 internal_trusted_storage or protected_storage.
+###############################################################################
+function(_create_psa_stdc_exe _exe_name _api_dir)
+
+	# Create the PSA test binary.
+	set(EXE_NAME ${_exe_name})
+
+	# Define PSA_STORAGE_LIB_NAME to be the name of the PSA Storage library to be tested.
```
Comment (jaypit02) [inline api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake]: You may want to guard this code for Crypto suite as this is related to storage tests.

Comment (simonqhughes) [reply api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake]: Guard added as requested, which I think is optimal because psa storage and psa crypto can be built alone, or they can also be tested in combination.

**Comment by simonqhughes** [comment]
@jaypit02 The new feedback has been acted upon. Please can you re-review and accept this PR? Thanks.

File: api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake [hunk 4]
```diff
@@ -15,6 +15,63 @@
...
+#                 internal_trusted_storage or protected_storage.
+###############################################################################
+function(_create_psa_stdc_exe _exe_name _api_dir)
+
+	# Create the PSA test binary.
+	set(EXE_NAME ${_exe_name})
+
+	if(DEFINED PSA_STORAGE_LIB_FILENAME)
+		# Define PSA_STORAGE_LIB_NAME to be the name of the PSA Storage library to be tested.
+		get_filename_component(PSA_STORAGE_LIB_NAME ${PSA_STORAGE_LIB_FILENAME} NAME)
```
Comment (gowthamsiddarthd) [inline api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake]: This doesn't work for libraries names with long extensions (.a.b.c  eg: libpsastorage.so.1.0.0)
My suggestion is to remove this line and add ${PSA_STORAGE_LIB_FILENAME} at line 68
This will allow the library names to be independent.

Same comments for crypto.

Comment (simonqhughes) [reply api-tests/platform/targets/tgt_dev_apis_stdc/target.cmake]: i dont understand this comment. I perform my testing using this Makefile:

https://github.com/ARMmbed/psa_trusted_storage_linux/blob/sdh_iotmbl_2565_psa_storage_building_test_binaries/test/Makefile

(The above file isnt on master because I need this PR merged first). At lines 95-97 in the above file you have these line:

```
# PSA_STORAGE_LIB_FILENAME is the location of the installed
# psa_trusted_storage_linux library
PSA_STORAGE_LIB_FILENAME=$(TOPDIR)/$(PSA_STORAGE_DIR)/$(PREFIX)/lib/libpsastorage.so
```
If you change line 97 to the following:

```
PSA_STORAGE_LIB_FILENAME=$(TOPDIR)/$(PSA_STORAGE_DIR)/$(PREFIX)/lib/libpsastorage.so.1.0.0
```

Then everything works fine with the version qualified libpsstorage.so.1.0.0, which is contrary to your findings i.e. you say it doesn't work. If you think about it, why wouldn't this work? Its just a library with a different name. This suggests there is some other cause for the problem you've encountered other than the name of the library.


P

**Comment by kotegowder** [comment]
The fixed was shared. 
Merging the changes for now. Subsequent changes will be pushed by psa-arch-tests team.
